packages/core/src/shared/utilities/resourceCache.ts

```diff
@@ -0,0 +1,177 @@
+import { getLogger } from '../logger/logger'
+import { waitUntil } from '../utilities/timeoutUtils'
+
+/**
+ * args:
+ *  @member result: the actual resource type callers want to use
+ *  @member locked: readWriteLock, while the lock is acquired by one process, the other can't access to it until it's released by the previous
+ *  @member timestamp: used for determining the resource is stale or not
+ */
+interface Resource<V> {
+    result: V | undefined
+    locked: boolean
+    timestamp: number
+}
+
+/**
+ * GlobalStates schema, which is used for vscode global states deserialization, [globals.globalState#tryGet<T>] etc.
+ * The purpose of it is to allow devs to overload the resource into existing global key and no need to create a specific key for only this purpose.
+ */
+export interface GlobalStateSchema<V> {
+    resource: Resource<V>
+}
+
+const logger = getLogger()
+
+function now() {
+    return globals.clock.Date.now()
+}
+
+/**
+ * constructor:
+ *  @param key: global state key, which is used for globals.globalState#update, #tryGet etc.
+ *  @param expirationInMilli: cache expiration time in milli seconds
+ *  @param defaultValue: default value for the cache if the cache doesn't pre-exist in users' FS
+ *  @param waitUntilOption: waitUntil option for acquire lock
+ *
+ * methods:
+ *  @method resourceProvider: implementation needs to implement this method to obtain the latest resource either via network calls or FS read
+ *  @method getResource: obtain the resource from cache or pull the latest from the service if the cache either expires or doesn't exist
+ */
+export abstract class CachedResource<V> {
+    constructor(
+        private readonly key: globalKey,
+        private readonly expirationInMilli: number,
+        private readonly defaultValue: GlobalStateSchema<V>,
+        private readonly waitUntilOption: { timeout: number; interval: number; truthy: boolean }
+    ) {}
+
+    abstract resourceProvider(): Promise<V>
+
+    async getResource(): Promise<V> {
+        const cachedValue = await this.tryLoadResourceAndLock()
+        const resource = cachedValue?.resource
+
+        // If cache is still fresh, return cached result, otherwise pull latest from the service
+        if (cachedValue && resource && resource.result) {
+            const duration = now() - resource.timestamp
+            if (duration < this.expirationInMilli) {
+                logger.info(`cache hit, duration(%sms) is less than expiration(%sms)`, duration, this.expirationInMilli)
+                // release the lock
+                await this.releaseLock(resource, cachedValue)
+                return resource.result
+            } else {
+                logger.info(
+                    `cached value is stale, duration(%sms) is older than expiration(%sms), invoking service API to pull the latest response`,
+                    duration,
+                    this.expirationInMilli
+                )
+            }
+        }
```

Comment:
avoid redundant messages

```suggestion
        } else {
            logger.info(`cache miss, invoking service API to pull the latest response`)
        }
```

Reply:
oh yes, that's a valid point, thanks for catching this justin